### PR TITLE
Add juice_add_turn_server API

### PR DESCRIPTION
**Pull request by Sean-Der**
Allows TURN servers to be added post agent creation

### CODE REVIEW

File: src/agent.c [hunk 1]
```diff
@@ -547,6 +553,44 @@ int agent_add_remote_candidate(juice_agent_t *agent, const char *sdp) {
 	return ret;
 }
 
+int agent_add_turn_server(juice_agent_t *agent, juice_turn_server_t *turn_server) {
+	int ret = -1;
+	conn_lock(agent);
```
Comment (paullouisageneau) [inline src/agent.c:558]: `conn_lock()` has no effect before gathering as there is no `conn_impl` created at this point, you should not call it.

File: src/agent.c [hunk 2]
```diff
@@ -547,6 +553,44 @@ int agent_add_remote_candidate(juice_agent_t *agent, const char *sdp) {
...
+	if (!agent_copy_turn_server(&agent->config.turn_servers[agent->config.turn_servers_count - 1],
+	                            turn_server)) {
+		goto exit;
+	}
+
+	ret = 0;
+
+exit:
+	conn_unlock(agent);
+	conn_interrupt(agent);
```
Comment (paullouisageneau) [inline src/agent.c:590]: You should not call `conn_unlock()` or `conn_interrupt()`, it has no effect as there is no `conn_impl`.

File: src/agent.c [hunk 3]
```diff
@@ -547,6 +553,44 @@ int agent_add_remote_candidate(juice_agent_t *agent, const char *sdp) {
...
+	}
+
+	if (!agent_copy_turn_server(&agent->config.turn_servers[agent->config.turn_servers_count - 1],
+	                            turn_server)) {
+		goto exit;
+	}
+
+	ret = 0;
+
+exit:
```
Comment (paullouisageneau) [inline src/agent.c:588]: You can clean up the label and simply `return -1` on error cases and 0 on success.

**Comment by Sean-Der** [comment]
Hey @paullouisageneau can I get another review please? I have addressed all the comments

thank you!

File: src/agent.c [hunk 4]
```diff
@@ -547,6 +553,31 @@ int agent_add_remote_candidate(juice_agent_t *agent, const char *sdp) {
 	return ret;
 }
 
+int agent_add_turn_server(juice_agent_t *agent, juice_turn_server_t *turn_server) {
```
Comment (paullouisageneau) [inline src/agent.c:556]:
```suggestion
int agent_add_turn_server(juice_agent_t *agent, const juice_turn_server_t *turn_server) {
```

File: src/agent.h [hunk 5]
```diff
@@ -158,6 +158,7 @@ int agent_resolve_servers(juice_agent_t *agent);
 int agent_get_local_description(juice_agent_t *agent, char *buffer, size_t size);
 int agent_set_remote_description(juice_agent_t *agent, const char *sdp);
 int agent_add_remote_candidate(juice_agent_t *agent, const char *sdp);
+int agent_add_turn_server(juice_agent_t *agent, juice_turn_server_t *turn_server);
```
Comment (paullouisageneau) [inline src/agent.h:161]:
```suggestion
int agent_add_turn_server(juice_agent_t *agent, const juice_turn_server_t *turn_server);
```

File: src/juice.c [hunk 6]
```diff
@@ -69,6 +69,16 @@ JUICE_EXPORT int juice_add_remote_candidate(juice_agent_t *agent, const char *sd
 	return JUICE_ERR_SUCCESS;
 }
 
+JUICE_EXPORT int juice_add_turn_server(juice_agent_t *agent, juice_turn_server_t *turn_server) {
```
Comment (paullouisageneau) [inline src/juice.c:72]:
```suggestion
JUICE_EXPORT int juice_add_turn_server(juice_agent_t *agent, const juice_turn_server_t *turn_server) {
```

File: include/juice/juice.h [hunk 7]
```diff
@@ -107,6 +107,7 @@ JUICE_EXPORT int juice_gather_candidates(juice_agent_t *agent);
 JUICE_EXPORT int juice_get_local_description(juice_agent_t *agent, char *buffer, size_t size);
 JUICE_EXPORT int juice_set_remote_description(juice_agent_t *agent, const char *sdp);
 JUICE_EXPORT int juice_add_remote_candidate(juice_agent_t *agent, const char *sdp);
+JUICE_EXPORT int juice_add_turn_server(juice_agent_t *agent, juice_turn_server_t *turn_server);
```
Comment (paullouisageneau) [inline include/juice/juice.h:110]:
```suggestion
JUICE_EXPORT int juice_add_turn_server(juice_agent_t *agent, const juice_turn_server_t *turn_server);
```

**Comment by Sean-Der** [comment]
@paullouisageneau can I get another review please! These two concerns are addressed/fixed

* Don't realloc/modify the existing turn_servers. I didn't feel comfortable handling all the realloc failure cases. Now I malloc and copy into it, if that fails the existing turn_servers aren't edited

* If the turn copy fails free the members

**Review by paullouisageneau — APPROVED** [review]
Looking good, thanks!

**Pull request merged**